index.d.ts

```diff
@@ -1,288 +1,530 @@
 declare module 'react-table' {
-  import { ReactNode, useState } from 'react'
+  import { ReactNode, useState, ComponentType, MouseEvent } from 'react'
+
+  export interface TableOptions<D extends object> extends UseTableOptions<D> {
+    //,
+    // UseSortByOptions<D>,
+    // UseFiltersOptions<D>,
+    // UseGroupByOptions<D>,
+    // UseExpandedOptions<D>,
+    // UsePaginationOptions<D>,
+    // UseRowSelectOptions<D>,
+    // UseRowStateOptions<D> {
+    // columns: [] // FIXME
+    // state: [] //FIXME
+  }
 
-  type StringKey<D> = Extract<keyof D, string>
-  type IdType<D> = StringKey<D> | string
+  export interface TableInstance<D extends object = {}>
+    extends TableOptions<D>,
+      UseTableInstanceProps<D> {}
 
-  type SortingRule<D> = {
-    id: IdType<D>
-    desc: boolean
-  }
+  /**
+   * The empty definition of TableState provides a definition for the state, that can then be extended in the users code.
+   *
+   * @example
+   *  export interface TableState<D extends object = {}}>
+   *    extends UsePaginationState,
+   *      UseGroupByState,
+   *      UseSortbyState<D>,
+   *      UseFiltersState<D> {}
+   */
+  export interface TableState<D extends object = {}> {} // eslint-disable-line @typescript-eslint/no-empty-interface
 
-  export type SortingRules<D> = SortingRule<D>[]
+  export type Hooks<D extends object = {}> = UseTableHooks<D> //&
+  // UseSortByHooks<D> &
+  // UseGroupByHooks<D> &
+  // UseExpandedHooks<D> &
+  // UseRowSelectHooks<D>
 
-  export type SortByFn = (a: any, b: any, desc: boolean) => 0 | 1 | -1
+  export type Cell<D extends object = {}> = UseTableCellProps<D>
 
-  export type Filters<D> = Record<IdType<D>, string>
+  export type Column<D extends object = {}> = UseTableColumnOptions<D>
```

Comment:
Column and ColumnInstance need to be interfaces so that you can do declaration merging on them, it isn't supported on types.

Reply:
probably Hooks does too.